abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+# Proxy-Wasm vNEXT ABI specification
+
+# Callbacks and Functions
+
+### Callbacks optionally exposed by the Wasm module
+
+All callbacks (entry points named `proxy_on_<event>`) are optional, and will be
+called only if exposed by the Wasm module.
+
+All callbacks, other than the integration functions, include unique context
+identifier as the first parameter, which can be used to distinguish between
+different contexts.
+
+### Functions exposed by the host environment
+
+All functions exposed by the host environment return `proxy_result_t`, which
+indicates the status of the call (success, invalid memory access, etc.), and
+return values are written into memory pointed by `return_<value>` parameters.
+
+
+## Integration
+
+### Functions exposed by the host environment
+
+#### `proxy_abi_version_X_Y_Z`
+
+* params:
+  - none
+* returns:
+  - none
+
+Exports ABI version (vX.Y.Z) that the Wasm module was compiled against.
+
+#### `proxy_log`
+
+* params:
+  - `i32 (proxy_log_level_t) log_level`
+  - `i32 (const char*) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Logs message (`message_data`, `message_size`) at the log level (`log_level`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `log_level`.
+
+Returns `InvalidMemoryAccess` when `message_data` and/or `message_size` point to
+invalid memory address.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (void*) memory_data`
+
+Allocates continuous memory buffer of size `memory_size` using the in-VM memory
+allocator and returns address of the memory buffer to the host environment using
+`memory_data`. Returning `0` indicates failure.
+
+
+## Context lifecycle
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+  - `i32 (proxy_context_type_t) context_type`
+* returns:
+  - `i32 (uint32_t) new_context_id`
+
+Called when the host environment creates a new context (`context_id`) of type
+`context_type` (e.g. `VmContext`, `PluginContext`, `StreamContext`,
```

Comment:
Is this spec supposed to cover `Access Logger` extensions ?

Do `Access Logger` extensions have a dedicated context type ?

Reply:
> No, since logging can happen in the "finalize" calls.

Handling access logs typically involves buffering, timeout, http/grpc call. All these actions have less sense in the context of an individual HTTP request / TCP connection than in the context of a dedicated `Access Logger` extension.